[WIP] Add copy functionality

amyrlam opened the pull request:
~~Has [LiveReload WIP commits](https://github.com/amyrlam/Docusaurus/pull/2) for ease of development~~

amyrlam committed:
Install clipboard.js

re: https://clipboardjs.com/

amyrlam committed:
Same as https://github.com/amyrlam/Docusaurus/pull/1/files, doesn't work

amyrlam committed:
Port of https://github.com/babel/website/pull/1594 to Docusaurus

This works but needs some CSS-ing and there is some extra stuff

yangshun commented:
This is awesome! It works at the moment, but as you said, some styling improvements need to be made. I have some suggestions for structural improvements as well.

Since we want to ship this to all Docusaurus users, it should be part of the core folder, within lib. This PR is currently adding the functionality to Docusaurus' website only, not the Docusaurus library. Try shifting the code into the library, within `lib`. We can create a new directory called `scripts` or  and add the `code-block-buttons.js` file in there instead. Also, we will need to tweak the server-side code to serve that file during development and generate.

Try to separate the LiveReload commits out as we want to merge this PR independently from LiveReload.

yangshun commented on website/siteConfig.js:
```diff
@@ -180,7 +180,11 @@ const siteConfig = {
-  scripts: ['https://buttons.github.io/buttons.js'],
+  scripts: [
+    'https://buttons.github.io/buttons.js',
+    'https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.0/clipboard.min.js',
```
These two scripts have to be shifted into either the default `siteConfig` or into the HTML of the base body structure. Putting it here only affects Docusaurus' website, not its users.

amyrlam replied:
That makes sense. I kept these here and copied them to `examples/basics/siteConfig.js`, following buttons.github.io. Is that what you meant?

yangshun commented on website/static/css/custom.css:
```diff
@@ -5,3 +5,47 @@
+  font-size: 11px;
+}
+
+.btn-open-in-repl {
```
Not needed

amyrlam replied:
Fixed, sorry I shouldn't have left all of this in, just wanted to post what I had functional at EOD.

yangshun commented on website/static/css/custom.css:
```diff
@@ -5,3 +5,47 @@
 table td:first-child > code {
   white-space: nowrap;
 }
+
```
These CSS should be part of `main.css` as well so as to affect Docusaurus users.

amyrlam replied:
Fixed, I moved this to `lib/static/css/main.css`.

yangshun commented on website/static/js/code-blocks-buttons.js:
```diff
@@ -0,0 +1,74 @@
+    button("Copy", "Copy code to clipboard", copyIcon, "btn-clipboard")
+  );
+
+  const tryIcon = '<svg xmlns="http://www.w3.org/2000/svg" width="12" width="12" viewBox="0 0 64 64"><path fill="currentColor" d="M32 64C14.4 64 0 49.6 0 32S14.4 0 32 0s32 14.4 32 32-14.4 32-32 32zm0-58.7C17.3 5.3 5.3 17.3 5.3 32s12 26.7 26.7 26.7 26.7-12 26.7-26.7S46.7 5.3 32 5.3z"/><path fill="currentColor" d="M42.1 31.1l-15.9-9.4c-1.2-.7-2.1-.2-2.1 1.2v18.3c0 1.4 1 1.9 2.1 1.2L42.1 33s.6-.4.6- ...
```
This is for Babel website's playground button, and is not needed here.

amyrlam replied:
👍

yangshun commented on website/static/js/code-blocks-buttons.js:
```diff
@@ -0,0 +1,74 @@
+  });
+
+  // not sure what this is
+  const presets = ["es2015", "stage-0", "react"];
```
This are babel presets, also used only for the playground.

amyrlam replied:
👍

yangshun reviewed (commented):
Please remove the code that is used only for Babel Playground. Try to understand what the code is doing and remove where unnecessary.

amyrlam committed:
Remove, was a test from WIP commit

amyrlam committed:
Add clipboard scripts to `examples/basics/siteConfig.js` to match buttons.github.io script

amyrlam committed:
Get rid of Try/repl stuff

amyrlam committed:
Move Copy and restyle

Move Copy inside of code block. The only "copy a code block" example I could find off the top of my head was this one: https://ember-learn.github.io/ember-cli-addon-docs/latest/docs/components/docs-viewer

Keep color yellow on hover

amyrlam committed:
Move CSS to main.css so available to all users

amyrlam committed:
Move code-block-buttons to lib/

This seems to still work without other changes? Confused.

amyrlam commented:
Removed the LiveReload stuff.

//

(1)
Should the directory for `code-block-buttons.js` be `static/js` or `scripts`? 

(2)
> Also, we will need to tweak the server-side code to serve that file during development and generate.

It seems to work with no changes? Confused why `js/code-blocks-buttons` works before and after [this commit](https://github.com/amyrlam/Docusaurus/pull/3/commits/b8210ea2396401eeda9adf926388a9a5988903bd)?

//

Let me know, can make more changes tomorrow.

//

As for the styling itself, now it's within the code block, no color change on highlight. 

I think it would look better if the blue outline on click is flush with the edges of the code block. But figured I'd just show this to you first to see if even a viable design. Since I suppose I would just measure that with "guess and check" in the browser.

![copy](https://user-images.githubusercontent.com/1372946/39416583-f67f187a-4c02-11e8-9506-2af12f3e6ae5.gif)

amyrlam committed:
fix rebase

amyrlam committed:
Tweak button styling

Now, blue highlight on click is within the code block

amyrlam commented:
I think this is ready for Docusaurus? I'm going to open a new PR on the main repo. I'll make any needed edits there. Sorry this one was so noisy.

yangshun commented:
Ok let's submit a PR and see. I just added preview builds last night. Excited to see it in action 😄 

Also test `docusaurus-init` locally and make sure it works for a fresh init of docusaurus.

amyrlam commented:
Ah ok. I did a fresh `docusaurus-init` in Docusaurus repo and it's not working. Debugging it...

Also, the color scheme to "match" with the dark code block on Docusaurus' blog rn will not be a good default

![copy](https://user-images.githubusercontent.com/1372946/39463158-2b9ae3c2-4ccb-11e8-893f-ff2d670baed7.gif)

amyrlam commented:
@yangshun Oh I guess that within Docusaurus, I'm still pinned to 1.0.14 Docusaurus version. 

On this branch I did a `rm -rf website`. Instead of https://github.com/facebook/Docusaurus/blob/master/website/package.json#L16, I'm linked to     `"docusaurus": "^1.0.14"`.

I manually changed that to `github:amyrlam/Docusaurus#amy/add-copy"`.

I then add 
```
```js
headerLinks: [
    ...
    { blog: true, label: 'Blog' },
    ...
]```
``` 
to a blog post, copy does not appear.

So I'm not sure what's wrong yet. Will further debug tomorrow. Maybe I have some code in the wrong folder, where it's not a part of the `docusaurus-init` process?

Is this the right way to test this locally? 

//

The default color of "Copy" will also have to be changed, so that it works with the Docusaurus default.

Our color scheme:

![copy-functionality](https://user-images.githubusercontent.com/1372946/39464445-2cbc4a32-4cd2-11e8-9da3-c97c2bfe6742.gif)

Default(?) color scheme (from docusaurus-starter-kit):

![image](https://user-images.githubusercontent.com/1372946/39464594-384efcfe-4cd3-11e8-8d93-fbd993212ddc.png)

Maybe just change to black.

amyrlam commented:
moving PR to main site with questions
